torch/distributions/lowrank_multivariate_normal.py

```diff
@@ -0,0 +1,181 @@
+        the small size "capacitance" matrix::
+            capacitance = I + scale_factor.T @ inv(scale_diag) @ scale_factor
+    """
+    arg_constraints = {"loc": constraints.real,
+                       "scale_factor": constraints.real,
+                       "scale_diag": constraints.positive}
+    support = constraints.real
+    has_rsample = True
+
+    def __init__(self, loc, scale_factor, scale_diag, validate_args=None):
+        if loc.dim() < 1:
+            loc = loc.unsqueeze(0)
+        event_shape = loc.shape[-1:]
+        if scale_factor.dim() < 2:
+            raise ValueError("scale_factor must be at least two-dimensional, "
+                             "with optional leading batch dimensions")
+        if scale_factor.shape[-2:-1] != event_shape:
+            raise ValueError("scale_factor must be a batch of matrices with shape {} x m"
+                             .format(event_shape[0]))
+        if scale_diag.shape[-1:] != event_shape:
+            raise ValueError("scale_diag must be a batch of vectors with shape {}".format(event_shape))
+
+        scale_batch_shape = _get_batch_shape(scale_factor, scale_diag)
+        try:
+            batch_shape = torch._C._infer_size(loc.shape[:-1], scale_batch_shape)
+        except RuntimeError:
+            raise ValueError("Incompatible batch shapes: loc {}, scale_factor {}, scale_diag {}"
+                             .format(loc.shape, scale_factor.shape, scale_diag.shape))
+
+        loc_shape = batch_shape + event_shape
+        self.loc = loc.expand(loc_shape)
+        self.scale_factor = scale_factor.expand(loc_shape + scale_factor.shape[-1:])
+        self.scale_diag = scale_diag.expand(loc_shape)
+        self._capacitance_tril = _batch_capacitance_tril(self.scale_factor, self.scale_diag)
+        super(LowRankMultivariateNormal, self).__init__(batch_shape, event_shape,
+                                                        validate_args=validate_args)
+
+    @property
+    def mean(self):
+        return self.loc
+
+    @property
+    def variance(self):
+        return self.scale_factor.pow(2).sum(-1) + self.scale_diag
+
+    @lazy_property
+    def scale_tril(self):
+        # The following identity is used to increase the numerically computation stability
+        # for Cholesky decomposition (see http://www.gaussianprocess.org/gpml/, Section 3.4.3):
+        #     W @ W.T + D = D1/2 @ (I + D-1/2 @ W @ W.T @ D-1/2) @ D1/2
+        # The matrix "I + D-1/2 @ W @ W.T @ D-1/2" has eigenvalues bounded from below by 1,
+        # hence it is well-conditioned and safe to take Cholesky decomposition.
+        n = self._event_shape[0]
+        scale_diag_sqrt_unsqueeze = self.scale_diag.sqrt().unsqueeze(-1)
+        Dinvsqrt_W = self.scale_factor / scale_diag_sqrt_unsqueeze
+        K = torch.matmul(Dinvsqrt_W, Dinvsqrt_W.transpose(-1, -2)).contiguous()
+        K.view(-1, n * n)[:, ::n + 1] += 1  # add identity matrix to K
```

Comment:
@apaszke In lastest commit, I make K contiguous and add identity to a batch of matrices like this. Hope that this is fine.